hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndexWithSql.scala

```diff
@@ -103,4 +117,154 @@ class TestPartitionStatsIndexWithSql extends HoodieSparkSqlTestBase {
       )
     }
   }
+
+  test("Test partition stats index on string type field with insert and file pruning") {
+    withTempDir { tmp =>
+      val tableName = generateTableName
+      val tablePath = s"${tmp.getCanonicalPath}/$tableName"
+      // Create table with date type partition
+      spark.sql(
+        s"""
+           | create table $tableName (
+           |    ts BIGINT,
+           |    uuid STRING,
+           |    rider STRING,
+           |    driver STRING,
+           |    city STRING,
+           |    state STRING
+           |) using hudi
+           | options(
+           |    primaryKey ='uuid',
+           |    preCombineField = 'ts',
+           |    hoodie.metadata.index.partition.stats.enable = 'true',
+           |    hoodie.metadata.index.column.stats.column.list = 'rider'
+           |)
+           |PARTITIONED BY (state)
+           |location '$tablePath'
+         """.stripMargin
+      )
+
+      spark.sql(s"INSERT INTO $tableName VALUES (1695159649,'334e26e9-8355-45cc-97c6-c31daf0df330','rider-A','driver-K','san_francisco','california')")
+      spark.sql(s"INSERT INTO $tableName VALUES (1695332066,'7a84095f-737f-40bc-b62f-6b69664712d2','rider-B','driver-L','new york city','new york')")
+      spark.sql(s"INSERT INTO $tableName VALUES (1695516137,'e3cf430c-889d-4015-bc98-59bdce1e530c','rider-C','driver-P','houston','texas')")
+      spark.sql(s"INSERT INTO $tableName VALUES (1695516137,'3eeb61f7-c2b0-4636-99bd-5d7a5a1d2c04','rider-D','driver-M','princeton','new jersey')")
+      spark.sql(s"INSERT INTO $tableName VALUES (1695332066,'1dced545-862b-4ceb-8b43-d2a568f6616b','rider-E','driver-O','austin','texas')")
+      spark.sql(s"INSERT INTO $tableName VALUES (1695091554,'e96c4396-3fad-413a-a942-4cb36106d721','rider-F','driver-M','sunnyvale','california')")
+
+      // Validate partition_stats index exists
+      val metaClient = HoodieTableMetaClient.builder()
+        .setBasePath(tablePath)
+        .setConf(HoodieTestUtils.getDefaultStorageConf)
+        .build()
+      assertResult(tableName)(metaClient.getTableConfig.getTableName)
+      assertTrue(metaClient.getTableConfig.getMetadataPartitions.contains(PARTITION_STATS.getPartitionPath))
+
+      // Test pruning
+      spark.sql("set hoodie.metadata.enable=true")
+      spark.sql("set hoodie.enable.data.skipping=true")
+      checkAnswer(s"select uuid, rider, city, state from $tableName where rider > 'rider-D'")(
+        Seq("1dced545-862b-4ceb-8b43-d2a568f6616b", "rider-E", "austin", "texas"),
+        Seq("e96c4396-3fad-413a-a942-4cb36106d721", "rider-F", "sunnyvale", "california")
+      )
```

Comment:
need set `hoodie.fileIndex.dataSkippingFailureMode` strict mode, and for `texas` and `california` partition, only have one fgId, can we controll the file size write more file in one partition? the partition stats is a aggre metric, I think it's necessary for this case.

Reply:
> need set hoodie.fileIndex.dataSkippingFailureMode strict mode

done

> for texas and california partition, only have one fgId, can we controll the file size write more fgId in one partition?

yes, what I did is set `hoodie.parquet.small.file.limit=0` and then each insert into goes into separate filegroup, so i combined texas and caliornia inserts into. This is what it looks like:
<img width="1166" alt="Screenshot 2024-06-29 at 3 29 53 PM" src="https://github.com/apache/hudi/assets/16440354/ac21b7e1-dad4-4c59-9588-f305ef3381d5">